data/libs/EquipSet.lua

```diff
@@ -1,404 +1,643 @@
+-- Register an event listener function to be notified of changes to this ship's
+-- equipment loadout.
+---@param fun EquipSet.Listener
+function EquipSet:AddListener(fun)
+	table.insert(self.listeners, fun)
+end
+
+-- Method: RemoveListener
+--
+-- Remove a previously-registered event listener function.
+function EquipSet:RemoveListener(fun)
+	utils.remove_elem(self.listeners, fun)
+end
+
+-- Method: Install
+--
+-- Install an equipment item in the given slot or in free equipment volume.
 --
 -- Parameters:
 --
---  item - the equipment to install
---  num - the number of pieces to install. If nil, only one will be installed.
---  slot - the slot where to install the equipment. It will be checked against
---         the equipment itself, the method will return -1 if the slot isn't
---         valid. If nil, the default slot for the equipment will be used.
+--  equipment  - EquipType, an equipment item instance to install on this ship.
+--               Must be an instance, not a equipment item prototype. Cannot be
+--               currently installed on this or any other ship.
 --
--- Return:
+--  slotHandle - optional HullConfig.Slot, the slot to install the item into or
+--               nil if the item does not support slot mounting. If present,
+--               must be a slot returned from calling GetSlotHandle or a similar
+--               function on this EquipSet instance.
 --
---  installed - the number of pieces actually installed, or -1 if the specified
---              slot is not valid.
+-- Returns:
 --
-function EquipSet:Add(ship, item, num, slot)
-	num = num or 1
-	if not slot then
-		slot = item:GetDefaultSlot(ship)
-	elseif not item:IsValidSlot(slot, ship) then
-		return -1
-	end
-	assert(slot ~= "cargo", "Cargo slots for equipment are no longer valid")
+--  installed - boolean, indicates whether the item was successfully installed
+--
+---@param equipment EquipType
+---@param slotHandle HullConfig.Slot?
+---@return boolean
+function EquipSet:Install(equipment, slotHandle)
+	local slotId = self.idCache[slotHandle]
 
-	local added = self:__Add_NoCheck(item, num, slot)
-	if added == 0 then
-		return 0
-	end
-	local postinst_diff = added - item:Install(ship, added, slot)
-	if postinst_diff > 0 then
-		self:__Remove_NoCheck(item, postinst_diff, slot)
-		added = added-postinst_diff
+	if slotHandle then
+		if not slotId then
+			return false -- No such slot!
+		end
+
+		if self.installed[slotId] then
+			return false -- Slot already full!
+		end
+
+		if not self:CanInstallInSlot(slotHandle, equipment) then
+			return false -- Doesn't fit!
+		end
+
+		self.installed[slotId] = equipment
+		self.cache[equipment] = slotId
+	else
+		if not self:CanInstallLoose(equipment) then
+			return false
+		end
+
+		table.insert(self.installed, equipment)
+		self.cache[equipment] = self.cacheIndex
+		self.cacheIndex = self.cacheIndex + 1
 	end
-	if added > 0 then
-		self:__TriggerCallbacks(ship, slot)
+
+	equipment:OnInstall(self.ship, slotHandle)
```

Comment:
`equipment:OnRemove()` is called _before_ the equipment is removed (but before the ship is updated).
`equipment:OnInstall()` is called _after_ the equipment is installed (and also before the ship is updated).

Is this intended? There's little documentation for the intended usage of the `OnRemove` / `OnInstall` callbacks. 

A further note: should it be possible for an equipment item to prevent its own installation/removal? For example, a passenger cabin containing passengers. ie, have `OnRemove` / `OnInstall` return a boolean indicating success.

Reply:
This is intentional. Both methods are meant to be called while the equipment item is "connected" to the ship for consistency's sake - they're symmetrical.

I intentionally avoided items being able to prevent their own removal by returning a value from the callback. It "looks" simple, but _there are too many systems involved_. If it needs to validate removal, it needs to return a reason for the UI to tell the player it can't be removed. Rather than try to shoehorn in a bunch of enums and validation into the function that should just be concerned with cleaning up after an equipment item, (and then handling the case where we need to cancel installation/removal halfway through) it's simpler to add a `CanBeInstalled`/`CanBeRemoved` function later which returns a boolean and an error message that's used to determine whether we even call `EquipSet:Remove()` at all.

Reply:
Ok, just requires care by modders/future developers to use these APIs correct.  But they are (mostly) well-documented, so that's a good start :)

Closing.